src/main/kotlin/com/sourcegraph/cody/config/CodyAuthenticationManager.kt

```diff
@@ -83,74 +84,58 @@ class CodyAuthenticationManager(val project: Project) : Disposable {
   @CalledInAny fun getAccounts(): Set<CodyAccount> = accountManager.accounts
 
   @CalledInAny
-  fun getActiveAccountTier(): CompletableFuture<AccountTier?> =
-      getActiveAccountTier(forceRefresh = false)
-
-  @CalledInAny
-  fun getIsTokenInvalid(): CompletableFuture<Boolean?> = getIsTokenInvalid(forceRefresh = false)
-
-  /**
-   * User account type can change because users can renew or cancel their subscriptions at any time.
-   * Components which state depends on this property should be checking state of
-   * `getActiveAccountTier` in non-blocking way or listen for `AccountSettingChangeContext` events
-   * to update their state accordingly later.
-   */
-  @CalledInAny
-  private fun getActiveAccountTier(forceRefresh: Boolean): CompletableFuture<AccountTier?> {
-    activeAccountTier?.let { if (!forceRefresh) return it }
+  private fun getAuthenticationState(): AuthenticationState {
+    val previousIsTokenInvalid = isTokenInvalid?.getNow(null)
+    val previousTier = tier?.getNow(null)
+    val isTokenInvalidFuture = CompletableFuture<Boolean>()
+    val tierFuture = CompletableFuture<AccountTier>()
+    val authenticationState = AuthenticationState(tierFuture, isTokenInvalidFuture)
+    val account = getActiveAccount() ?: return authenticationState
+    val token = account.let(::getTokenForAccount)
 
-    val account = getActiveAccount() ?: return CompletableFuture.completedFuture(null)
-    val previousAccountTier = activeAccountTier?.getNow(null)
-    val accountTierFuture = CompletableFuture<AccountTier?>()
-    activeAccountTier = accountTierFuture
+    isTokenInvalid = isTokenInvalidFuture
+    tier = tierFuture
 
-    accountTierFuture.thenApply { currentAccountTier ->
-      if (previousAccountTier != currentAccountTier) {
+    tierFuture.thenApply { currentAccountTier ->
```

Comment:
Looks like we are publisher notification both in `getAuthenticationState` and `setActiveAccount`. Won't that cause some of the notification to be sent twice?
Somehow I'm not sure about that whole design we have there.